SnowplowTracker/SnowplowTracker/Emitter.cs

```diff
@@ -0,0 +1,484 @@
+		                int sendLimit = 500, int bufferSize = 25, long byteLimitGet = 52000, long byteLimitPost = 52000) {
+			Utils.CheckArgument (!String.IsNullOrEmpty (endpoint), "Endpoint cannot be null or empty.");
+			this.endpoint = endpoint;
+			this.collectorUri = MakeCollectorUri(endpoint, protocol, method);
+			this.httpProtocol = protocol;
+			this.httpMethod = method;
+			this.sendLimit = sendLimit;
+			this.bufferSize = bufferSize;
+			this.byteLimitGet = byteLimitGet;
+			this.byteLimitPost = byteLimitPost;
+			this.eventStore = new EventStore (emitLock);
+		}
+
+		/// <summary>
+		/// Adds an event payload to the database.
+		/// </summary>
+		/// <param name="payload">Payload.</param>
+		public void Add(TrackerPayload payload) {
+			this.eventStore.AddEvent (payload);
+
+		}
+
+		/// <summary>
+		/// Starts the Emitter; will be run every 10ms.
+		/// </summary>
+		public void Start() {
+			if (!sending) {
+				// Start the emitter sending loop.
+				emitThread = new Thread(new ThreadStart(Emit));
+				sending = true;
+				emitThread.Start ();
+
+				// Start the event consumer.
+				eventStore.StartConsumer();
+			}
+		}
+		
+		/// <summary>
+		/// Stops the Emitter; will also empty the queue into the database.
+		/// </summary>
+		public void Stop() {
+			if (sending) {
+				// Stop the emitter sending loop.
+				sending = false;
+				lock (emitLock) {
+					Monitor.Pulse(emitLock);
+				}
+
+				// Stops the event consumer.
+				eventStore.StopConsumer();
+			}
+		}
+	
+		/// <summary>
+		/// Performs and orchestrates all sending of events from the database
+		/// and subsequent removal of events from the database.
+		/// </summary>
+		private void Emit() {
+			while (sending) {
+				List<EventRow> events = eventStore.GetDescEventRange (sendLimit);
+
+				// If some events were found start sending!
+				if (events.Count > 0) {
+					Log.Debug ("Emitter: Event count: " + events.Count);
+					List<RequestResult> results = SendRequests (events);
+					events = null;
+					
+					int success = 0;
+					int failure = 0;
+					
+					List<int> eventsToDelete = new List<int>();
+					
+					foreach (RequestResult result in results) {
+						if (result.success) {
+							eventsToDelete.AddRange(result.rowIds);
+							success += result.rowIds.Count;
+						} else {
+							failure += result.rowIds.Count;
+						}
+					}
+					
+					eventStore.DeleteEvents(eventsToDelete);
+					
+					Log.Debug ("Emitter: event sending results.");
+					Log.Debug (" + Successful: " + success);
+					Log.Debug (" + Failure: " + failure);
+					
+					if (failure > 0 && success == 0) {
+						Log.Error ("Emitter: All events failed to send; pausing emitter for ten seconds.");
+						Thread.Sleep (FAIL_INTERVAL);
+					} else {
+						Log.Debug ("Emitter: All events sent successfully; checking for more...");
+					}
+				} else {
+					Log.Debug ("Emitter: No more events to send; waiting for more.");
+					lock (emitLock) {
+						Monitor.Wait(emitLock);
```

Comment:
I think the check that there are events in the event store needs to come inside the lock block. Otherwise the following sequence is possible:

The Emit method calls eventStore.GetDescEventRange and gets an empty list
Then a different thread adds events to the eventStore and Monitor.Pulse(emitLock) is called but the Emit thread is not waiting and misses the pulse
The Emit thread now reaches and enters the lock but has missed the pulse and goes to sleep indefinitely despite the fact that there are events in eventStore. Those events won't be sent until more events are added to the eventStore or the application is restarted.